### PR TITLE
feat: Improve completion with snippet support (#786)

### LINKED ISSUE #786: Improve completion with snippet support
Today completion for Qute and application.properties doesn't support snippet.

We should support it.

### LINKED ISSUE #975: Auto-completion of Qute tags
when using auto-completion of `{#if` for example, the closing tag is added with wrong indent:
image

I think it would be better to keep the snippet on one line and let the user decide of the rest.

### PR DESCRIPTION
**Pull request by angelozerr**
feat: Improve completion with snippet support (#786)

Fixes #786
Fixes #975

### CODE REVIEW

**Comment by angelozerr** [comment]
Snippet completion starts working:

* here a demo with Qute template

SnippetCompletionWithQute

//cc @ia3andy 

 * here a demo with properties in application.properties:
 
SnippetCompletionWithAppProp

//cc @gastaldi 

 * here a demo in Java file:

SnippetCompletioInJava

I need to clean my code

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/SnippetTemplateFactory.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion;
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/SnippetTemplateFactory.java:1]: header

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/SnippetTemplateLoader.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+public class SnippetTemplateLoader implements LspSnippetHandler  {
+
+    private final Template template;
+
+    private final Function<String, String> variableResolvers;
+
+    private final List<String> existingVariables;
+
+    public SnippetTemplateLoader(Template template, Function<String, String> variableResolvers) {
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/SnippetTemplateLoader.java:23]: variableResolver

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/SnippetTemplateLoader.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion;
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/SnippetTemplateLoader.java:1]: header

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/Location.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet;
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/Location.java:1]: header

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/LspSnippetHandler.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet;
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/LspSnippetHandler.java:1]: header

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/LspVariables.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet;
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/LspVariables.java:1]: header

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/ParseException.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet;
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/ParseException.java:1]: header

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/ParseException.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+   * Returns the column number at which the error occurred, i.e. the number of the character in its
+   * line. The number of the first character of a line is 1.
+   *
+   * @return the column in which the error occurred, will be &gt;= 1
+   * @deprecated Use {@link #getLocation()} instead
+   */
+  @Deprecated
+  public int getColumn() {
+    return location.column;
+  }
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/ParseException.java:58]: no point in adding `@Deprecated` methods in a new file. Either remove the methods or remove the annotation.

File: src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/AdvancedTest.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet;
```
Comment (fbricon) [inline src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/AdvancedTest.java:1]: header

File: src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/ChoiceTest.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet;
```
Comment (fbricon) [inline src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/ChoiceTest.java:1]: header

File: src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/LspSnippetAssert.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet;
```
Comment (fbricon) [inline src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/LspSnippetAssert.java:1]: header

File: src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/PlaceholderTest.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet;
```
Comment (fbricon) [inline src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/PlaceholderTest.java:1]: header

File: src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/TabstopTest.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet;
```
Comment (fbricon) [inline src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/TabstopTest.java:1]: header

File: src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/TextTest.java [hunk 14]
```diff
@@ -0,0 +1,19 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet;
```
Comment (fbricon) [inline src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/TextTest.java:1]: header

File: src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/VariableTest.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet;
```
Comment (fbricon) [inline src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/VariableTest.java:1]: header

File: src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/handler/LspSnippetHandlerImpl.java [hunk 16]
```diff
@@ -0,0 +1,70 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet.handler;
```
Comment (fbricon) [inline src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/handler/LspSnippetHandlerImpl.java:1]: header

File: src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/handler/LspSnippetNode.java [hunk 17]
```diff
@@ -0,0 +1,5 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet.handler;
```
Comment (fbricon) [inline src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/handler/LspSnippetNode.java:1]: header

File: src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/handler/TabstopNode.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet.handler;
```
Comment (fbricon) [inline src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/handler/TabstopNode.java:1]: header

**Review by fbricon — CHANGES_REQUESTED** [review]

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/LspSnippetHandler.java (hunk 5 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/LspSnippetHandler.java:14]: fixed

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/ParseException.java (hunk 7 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/ParseException.java:22]: fixed

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/SnippetTemplateFactory.java [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+    /**
+     * Returns an Intellij Template instanced loaded from the given LSP snippet content.
+     *
+     * @param snippetContent    the LSP snippet content
+     * @param project           the project.
+     * @param variableResolvers the LSP variable resolvers (ex to resolve ${TM_SELECTED_TEXT}).
+     * @return an Intellij Template instanced loaded from the given LSP snippet content.
+     */
+    public static @NotNull Template createTemplate(@NotNull String snippetContent, @NotNull Project project, @NotNull Function<String, String> variableResolvers) {
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/SnippetTemplateFactory.java:40]: variableResolver

File: src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/handler/TextNode.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
+package com.redhat.devtools.intellij.lsp4ij.operations.completion.snippet.handler;
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/test/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/snippet/handler/TextNode.java:14]: fixed

**Review by fbricon — APPROVED** [review]

**Pull request merged**